### PR TITLE
Add submission Meta Attributes feature.

### PR DESCRIPTION
**Pull request by KhalidRmb**
ToDo:
- [x]  Finish dynamic form for taking the attributes.

### CODE REVIEW

File: apps/challenges/views.py [hunk 1]
```diff
@@ -808,6 +808,47 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
                 response_data, status=status.HTTP_400_BAD_REQUEST
             )
 
+        # Move this to another method.
+        if(data.get("submission_meta_attributes")):  # To ensure that the schema for submission metaattributes is valid.
```
Comment (Ram81) [inline apps/challenges/views.py:812]: @KhalidRmb there is another API for validate challenge config we have to add these changes to that  API too. It will be used for github integration of challenge creation flow.

File: apps/challenges/views.py [hunk 2]
```diff
@@ -814,6 +814,46 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
...
+                    name = attribute["name"]
+                except KeyError:
+                    message = (
+                    "No name found for attribute in challenge_phase {}".format(data["id"])
+                    )
+                    response_data = {"error": message}
+                    return Response(response_data, status=status.HTTP_406_NOT_ACCEPTABLE)
+
+                try:
+                    description = attribute["description"]
```
Comment (Ram81) [inline apps/challenges/views.py:830]: @KhalidRmb can we create a generic method in utils which will take json object and `key` and will return whether that key exists or not. That way code will be clean and you won't need multiple `try except`

File: apps/challenges/views.py [hunk 3]
```diff
@@ -814,6 +814,46 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
             response_data = {"error": message}
             return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
 
+        if(data.get("submission_meta_attributes_schema")):  # To ensure that the schema for submission metaattributes is valid.
```
Comment (Ram81) [inline apps/challenges/views.py:817]: add space after `if` and move comment above if condition

File: apps/jobs/models.py [hunk 4]
```diff
@@ -116,6 +116,7 @@ class Submission(TimeStampedModel):
         blank=True,
         null=True,
     )
+    submission_meta_attributes = JSONField(blank=True, null=True) #  Store the values of meta attributes for the submission here.
```
Comment (Ram81) [inline apps/jobs/models.py:119]: move comment above field

File: frontend/src/views/web/challenge/submission-meta-attributes-dialog.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    <div class="row" ng-repeat="a in challenge.meta_attributes_data">
+        <div class="col s12 m12" ng-if="a.value!=undefined">
+            <strong>{{a.name}}</strong> : {{a.value}}
+        </div>
+        <div class="col s12 m12" ng-if="a.values!=undefined">
+            <strong>{{a.name}}</strong> : <span class="col s12 m12" ng-repeat="value in a.values">{{value}}</span>
+        </div>
+    </div>
+    <button class="btn ev-btn-dark waves-effect waves-dark grad-btn grad-btn-dark fs-14" type="cancel" ng-click="challenge.showMetaAttributesDialog($event, false);">Close</button>
+</section>
```
Comment (Ram81) [inline frontend/src/views/web/challenge/submission-meta-attributes-dialog.html:11]: add new line

File: frontend/src/js/controllers/challengeCtrl.js [hunk 6]
```diff
@@ -451,6 +453,9 @@
                     formData.append("method_description", vm.methodDesc);
                     formData.append("project_url", vm.projectUrl);
                     formData.append("publication_url", vm.publicationUrl);
+                    formData.append("submission_meta_attributes", vm.submission_meta_attributes);
```
Comment (KhalidRmb) [inline frontend/src/js/controllers/challengeCtrl.js:456]: @Ram81 This is where I'm storing the attributes. Do I need to convert it into something before appending here. It should be stored as JSONField on the backend.

**Comment by Ram81** [comment]
@KhalidRmb can you fix the travis build?

**Comment by KhalidRmb** [comment]
> @KhalidRmb can you fix the travis build?

Hey, so a lot of tests are failing because I've changed the challenge phase to include a new field which can be blank. Since many tests assert the equivalence of a challenge phase object with an expected one which is missing the new attribute, those tests are failing. 

Any way to bypass this without changing every one of those tests?

**Comment by Ram81** [comment]
@KhalidRmb I don't think so. But it is a good idea to have new fields added so that you can verify your serializer functionality.

File: frontend/src/js/controllers/challengeCtrl.js [hunk 7]
```diff
@@ -451,6 +453,9 @@
                     formData.append("method_description", vm.methodDesc);
                     formData.append("project_url", vm.projectUrl);
                     formData.append("publication_url", vm.publicationUrl);
+                    formData.append("submission_meta_attributes", JSON.stringify(vm.submission_meta_attributes));
+                    console.log("While making submission, vm.submission_meta_attributes is:")
+                    console.log(vm.submission_meta_attributes)
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:458]: Can we remove these? I believe you were using these for debugging

File: frontend/src/js/controllers/challengeCtrl.js [hunk 8]
```diff
@@ -508,7 +514,9 @@
             }
         };
 
-
+        vm.consoler = function(){
+            console.log("LOLOLOLOLOL");
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:518]: same here

File: frontend/src/js/controllers/challengeCtrl.js [hunk 9]
```diff
@@ -518,6 +526,7 @@
             onSuccess: function(response) {
                 var details = response.data;
                 vm.phases = details;
+                console.log(details);
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:529]: same here

File: frontend/src/css/modules/submission.scss [hunk 10]
```diff
@@ -63,3 +63,25 @@ code {
...
+.submission-options {
+    margin-left: 10px;
+}
+md-radio-button.md-default-theme .md-on, md-radio-button .md-on {
+    background-color: black;
+}
+md-radio-button.md-default-theme.md-checked .md-off, md-radio-button.md-checked .md-off {
+    border-color: black;
+}
+md-radio-button .md-off {
```
Comment (Ram81) [inline frontend/src/css/modules/submission.scss:76]: add new line above every new css class

File: frontend/src/css/modules/submission.scss [hunk 11]
```diff
@@ -63,3 +63,25 @@ code {
...
+    height: 16px;
+}
+md-radio-button .md-on {
+    width: 16px;
+    height: 16px;
+}
+[type="checkbox"] + label:before, [type="checkbox"]:not(.filled-in) + label:after {
+    width: 16px;
+    height: 16px;
+}
```
Comment (Ram81) [inline frontend/src/css/modules/submission.scss:87]: add new line here

File: apps/challenges/utils.py [hunk 12]
```diff
@@ -32,6 +32,13 @@
 get_challenge_phase_split_model = get_model_object(ChallengePhaseSplit)
 
 
+def key_exists_in_json(json, key):
+    try:
+        name = json[key]
```
Comment (Ram81) [inline apps/challenges/utils.py:37]: this can be done in better way by getting `json.keys()` and using `key in keys`

Comment (KhalidRmb) [reply apps/challenges/utils.py:37]: That way...I guess I'll send all the missing keys at once in a message. The method could return a list.

**Comment by RishabhJain2018** [comment]
@KhalidRmb Can you please resolve the conflicts here.

File: apps/challenges/views.py [hunk 13]
```diff
@@ -816,6 +817,28 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
             response_data = {"error": message}
             return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
 
+        # To ensure that the schema for submission metaattributes is valid.
+        if (data.get("submission_meta_attributes_schema")):
+            for attribute in data["submission_meta_attributes_schema"]:
+                keys = ["name", "description", "type"]
+                missing = keys_exists_in_json(attribute, keys)
+
+                if missing == []:
```
Comment (Ram81) [inline apps/challenges/views.py:826]: do a length check `len(missing) == 0` the way you did is not the right way to do it

File: apps/jobs/migrations/0017_add_submission_meta_attributes_field_to_jobs.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+from __future__ import unicode_literals
+
+import django.contrib.postgres.fields.jsonb
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('jobs', '0016_add_ignore_submission_boolean_field_in_submission_model'),
```
Comment (Ram81) [inline apps/jobs/migrations/0017_add_submission_meta_attributes_field_to_jobs.py:12]: is this also related to submission meta attributes? unclear to me how ignore_submission is related to that

File: frontend/src/js/controllers/challengeCtrl.js [hunk 15]
```diff
@@ -603,11 +620,41 @@
...
+            }).attributes;
+        };
+
+        vm.clearMetaAttributeValues = function(){
+            if(vm.submissionMetaAttributes != null){
+                vm.submissionMetaAttributes.forEach(function(attribute){
+                    if(attribute.type == 'checkbox'){
+                        attribute.values = [];
+                    }
+                    else{
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:635]: add space after both `if` and `else`

File: apps/jobs/migrations/0017_add_submission_meta_attributes_field_to_jobs.py (hunk 14 above)
Comment (KhalidRmb) [reply apps/jobs/migrations/0017_add_submission_meta_attributes_field_to_jobs.py:12]: It's a dependency. It's not related.

File: apps/challenges/aws_utils.py [hunk 16]
```diff
@@ -579,6 +603,18 @@ def scale_workers(queryset, num_of_tasks):
     dict: keys-> 'count': the number of workers successfully started.
                  'failures': a dict of all the failures with their error messages and the challenge pk
     """
+    if ENV == 'dev':
+        logger.info("Entered scale_workers method.")
```
[the remark this answers is not in the document]
Comment (KhalidRmb) [reply apps/challenges/aws_utils.py:607]: They are just some error catching changes for dev environments. I made them while working on this, and would be quicker to get merged. Should I make a separate PR?

File: apps/challenges/migrations/0067_add_sub_meta_att_schema_to_challenge_phase.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+from __future__ import unicode_literals
+
+import django.contrib.postgres.fields.jsonb
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('challenges', '0066_added_is_partial_submission_evaluation_enabled_field_in_phase_model'),
```
[the remark this answers is not in the document]
Comment (KhalidRmb) [reply apps/challenges/migrations/0067_add_sub_meta_att_schema_to_challenge_phase.py:12]: Another dependency right.

File: apps/challenges/aws_utils.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Ram81) [reply apps/challenges/aws_utils.py:607]: sure, let them be in the same PR then

File: apps/challenges/views.py [hunk 18]
```diff
@@ -816,6 +817,28 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
...
+                    if attribute_type in valid_attribute_types:
+                        if attribute_type == "radio" or attribute_type == "checkbox":
+                            options = attribute.get("options")
+                            if not options or not len(options):
+                                message = "Please include at least one option in attribute for challenge_phase {}".format(data["id"])
+                                response_data = {"error": message}
+                                return Response(response_data, status=status.HTTP_406_NOT_ACCEPTABLE)
+                else:
+                    missing_string = ", ".join(missing)
+                    message = "Please enter the following to attribute in phase {}: {}.".format(data["id"], missing_string)
```
Comment (Ram81) [inline apps/challenges/views.py:838]: message isn't clear should it be `Please enter the following to meta attributes in phase`? or `Please enter the following to submission meta attributes in phase`?

File: frontend/src/js/controllers/challengeCtrl.js [hunk 19]
```diff
@@ -747,6 +794,13 @@
                         }
                     }
                     for (var i=0; i<vm.leaderboard.length; i++) {
+                        if (vm.leaderboard[i].submission__submission_meta_attributes == null){
+                            vm.IsLeaderboardHavingMetaAtts = false;
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:798]: replace with `vm.isLeaderboardHavingMetaAttributes`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 20]
```diff
@@ -826,6 +880,32 @@
             vm.getLeaderboard(vm.phaseSplitId);
         }
 
+        
+        vm.showMetaAttributesDialog = function(ev, attributes){
+            if (attributes != false){
+                vm.meta_attributes_data = [];
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:886]: use CamelCase, please be consistent

File: apps/challenges/views.py (hunk 18 above)
Comment (KhalidRmb) [reply apps/challenges/views.py:838]: Will make the change.

File: frontend/src/js/controllers/challengeCtrl.js (hunk 19 above)
Comment (Ram81) [reply frontend/src/js/controllers/challengeCtrl.js:798]: @KhalidRmb change this to `showSubmissionMetaAttributesOnLeaderboard` or something relevant. `isLeaderboardHavingMetaAttributes` doesn't sound right to me

Comment (KhalidRmb) [reply frontend/src/js/controllers/challengeCtrl.js]: Done.

**Comment by KhalidRmb** [comment]
@RishabhJain2018 PTAL.

File: apps/challenges/aws_utils.py [hunk 21]
```diff
@@ -747,20 +812,22 @@ def get_logs_from_cloudwatch(log_group_name, log_stream_prefix, start_time, end_
...
-            filterPattern=pattern
-        )
-        for event in response["events"]:
-            logs.append(event["message"])
-    except ClientError as e:
-        logger.exception(e)
-        return ["There was an error in displaying the logs."]
-
+    if ENV == 'dev':
+        logs = ["Sample Log. Can't fetch Cloudwatch logs in test/dev environment."]
```
Comment (RishabhJain2018) [inline apps/challenges/aws_utils.py:816]: `["Sample Log. Can't fetch Cloudwatch logs in test/dev environment."]` -> `["The worker logs in the development environment are available on the terminal. Please use docker-compose worker  -f to view the logs."]`

File: apps/challenges/aws_utils.py [hunk 22]
```diff
@@ -747,20 +812,22 @@ def get_logs_from_cloudwatch(log_group_name, log_stream_prefix, start_time, end_
...
+                logStreamNamePrefix=log_stream_prefix,
+                startTime=start_time,
+                endTime=end_time,
+                filterPattern=pattern
+            )
+            for event in response["events"]:
+                logs.append(event["message"])
+        except Exception as e:
+            logger.exception(e)
+            return ["There was an error in displaying the logs."]
```
Comment (RishabhJain2018) [inline apps/challenges/aws_utils.py:830]: `[f"There is an error in displaying logs. Please find the full error traceback here {e}"]`

File: apps/challenges/challenge_notification_util.py [hunk 23]
```diff
@@ -1,13 +1,21 @@
...
+import os
 
 from django.conf import settings
 
 from base.utils import send_email
 
 logger = logging.getLogger(__name__)
 
+DJANGO_SETTINGS_MODULE = os.environ.get("DJANGO_SETTINGS_MODULE")
+ENV = DJANGO_SETTINGS_MODULE.split(".")[-1]
```
Comment (RishabhJain2018) [inline apps/challenges/challenge_notification_util.py:11]: Can we please remove this variable.

File: apps/challenges/challenge_notification_util.py [hunk 24]
```diff
@@ -1,13 +1,21 @@
...
+
 
 def construct_and_send_worker_start_mail(challenge):
+    if ENV == "dev":
```
Comment (RishabhJain2018) [inline apps/challenges/challenge_notification_util.py:15]: Same as above.

File: apps/challenges/challenge_notification_util.py [hunk 25]
```diff
@@ -1,13 +1,21 @@
...
+        logger.info("Inside construct_and_send_worker_start_mail method.")
```
Comment (RishabhJain2018) [inline apps/challenges/challenge_notification_util.py:16]: Can we please remove this debugging statement.

File: apps/challenges/migrations/0067_add_sub_meta_att_schema_to_challenge_phase.py [hunk 26]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
```
Comment (RishabhJain2018) [inline apps/challenges/migrations/0067_add_sub_meta_att_schema_to_challenge_phase.py:1]: Can we please change the filename to `0067_add_sub_meta_att_schema_to_challenge_phase.py` -> `0067_add_submission_meta_attribute_field_to_challenge_phase.py`

File: apps/challenges/models.py [hunk 27]
```diff
@@ -256,6 +256,8 @@ def __init__(self, *args, **kwargs):
     )
     # Flag to restrict user to select only one submission for leaderboard
     is_restricted_to_select_one_submission = models.BooleanField(default=False)
+    # Store the schema for the submission meta attributes of this challenge phase.
+    submission_meta_attributes_schema = JSONField(default=None, blank=True, null=True)
```
Comment (RishabhJain2018) [inline apps/challenges/models.py:260]: `submission_meta_attributes_schema` -> `submission_meta_attributes`

File: apps/challenges/serializers.py [hunk 28]
```diff
@@ -92,6 +92,7 @@ class Meta:
             "slug",
             "max_concurrent_submissions_allowed",
             "is_restricted_to_select_one_submission",
+            "submission_meta_attributes_schema",
```
Comment (RishabhJain2018) [inline apps/challenges/serializers.py:95]: `submission_meta_attributes`

File: apps/challenges/serializers.py [hunk 29]
```diff
@@ -279,6 +280,7 @@ class Meta:
             "max_concurrent_submissions_allowed",
             "environment_image",
             "is_restricted_to_select_one_submission",
+            "submission_meta_attributes_schema",
```
Comment (RishabhJain2018) [inline apps/challenges/serializers.py:283]: `submission_meta_attributes`

File: apps/challenges/views.py [hunk 30]
```diff
@@ -816,6 +817,28 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
             response_data = {"error": message}
             return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
 
+        # To ensure that the schema for submission metaattributes is valid.
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:820]: `# To ensure that the schema for submission metaattributes is valid.` -> `# To ensure that the schema for submission meta attributes is valid.`

File: apps/challenges/utils.py [hunk 31]
```diff
@@ -32,6 +32,14 @@
 get_challenge_phase_split_model = get_model_object(ChallengePhaseSplit)
 
 
+def keys_exists_in_json(json, keys):
```
Comment (RishabhJain2018) [inline apps/challenges/utils.py:35]: def get_missing_keys_from_json(json, keys):

File: apps/challenges/utils.py [hunk 32]
```diff
@@ -32,6 +32,14 @@
...
+    missing = []
```
Comment (RishabhJain2018) [inline apps/challenges/utils.py:36]: Can we please write down the docstrings as well.

File: apps/challenges/views.py [hunk 33]
```diff
@@ -816,6 +817,28 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
...
+        if (data.get("submission_meta_attributes_schema")):
+            for attribute in data["submission_meta_attributes_schema"]:
+                keys = ["name", "description", "type"]
+                missing = keys_exists_in_json(attribute, keys)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:824]: `missing` -> `missing_keys`

File: apps/jobs/models.py [hunk 34]
```diff
@@ -118,6 +118,8 @@ class Submission(TimeStampedModel):
         blank=True,
         null=True,
     )
+    # Store the values of meta attributes for the submission here.
+    submission_meta_attributes = JSONField(blank=True, null=True)
```
Comment (RishabhJain2018) [inline apps/jobs/models.py:122]: `submission_meta_attributes` -> `submission_metadata`

File: apps/jobs/serializers.py [hunk 35]
```diff
@@ -53,6 +53,7 @@ class Meta:
             "when_made_public",
             "is_baseline",
             "job_name",
+            "submission_meta_attributes"
```
Comment (RishabhJain2018) [inline apps/jobs/serializers.py:56]: `submission_metadata`

File: apps/jobs/serializers.py [hunk 36]
```diff
@@ -121,6 +122,7 @@ class Meta:
             "created_at",
             "method_name",
             "participant_team_members",
+            "submission_meta_attributes"
```
Comment (RishabhJain2018) [inline apps/jobs/serializers.py:125]: `submission_metadata`

File: apps/jobs/utils.py [hunk 37]
```diff
@@ -321,6 +321,8 @@ def calculate_distinct_sorted_leaderboard_data(
         "leaderboard__schema",
         "submission__submitted_at",
         "submission__method_name",
+        "submission__id",
+        "submission__submission_meta_attributes"
```
Comment (RishabhJain2018) [inline apps/jobs/utils.py:325]: `submission__submission_meta_attributes` -> `submission__submission_metadata`

File: apps/jobs/views.py [hunk 38]
```diff
@@ -272,6 +272,13 @@ def challenge_submission(request, challenge_id, challenge_phase_id):
                 "message": "Please wait while your submission being evaluated!"
             }
             return Response(response_data, status=status.HTTP_200_OK)
+
+        try:
+            submission_meta_attributes = json.load(request.data.get('submission_meta_attributes'))
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:277]: How about if we use `if request.data.get('submission_meta_attributes'):` rather than a try and logging.info running for every submission

**Comment by KhalidRmb** [comment]
@RishabhJain2018 Done!

**Comment by Ram81** [comment]
@RishabhJain2018 tested the PR. It works fine for me 👍
